src/core/lwt.ml

```diff
@@ -22,1307 +22,2777 @@
  * 02111-1307, USA.
  *)
 
-(* +-----------------------------------------------------------------+
-   | Types                                                           |
-   +-----------------------------------------------------------------+ *)
 
-exception Canceled
 
-module Int_map = Map.Make(struct type t = int let compare = compare end)
+(* Reading guide
 
-type data = (unit -> unit) Int_map.t
-  (* Type of data carried by threads *)
+   Welcome to the implementation of the Lwt core! This is a big file, but we
+   hope that reading it (parts at a time!) will not be scary :) Here is why:
 
-type +'a t
-type -'a u
 
-(* A type + a thread of this type. *)
-module type A_thread = sig
-  type a
-  val thread : a t
-end
+   * Sectioning
+
+   The code is broken up into sections, each one of which is an internal module.
+   Most of the modules have a signature, which serves as a neat table of
+   contents.
+
+   It is recommended that you read this file with code folding enabled. If you
+   fold all the modules, you can visualize the logical structure of Lwt quite
+   easily. You can then expand modules as needed, depending on what part of the
+   implementation you are interested in. Without code folding, you face an
+   intimidating wall of code :( You can still visually parse the file, however,
+   because there are plenty of blank lines to help section things off. You can
+   also view this file folded online:
+
+     https://gist.github.com/aantron/9fab0bdead98a60fccf06e0189186863
+     https://gist.github.com/aantron/97b58520d5bb4858ccac6f54700a24d7
+
+   The signatures are unusual: big comments are absent. They are moved into the
+   modules, so that they are hidden by code folding when you (the reader!) are
+   not interested in those modules.
+
+
+   * Documentation
+
+   The documentation begins with an Overview of major concepts and components.
```

Comment:
I actually wrote that deliberately, but looking back on it months later, I have to agree :) Changing...